### ISSUE TITLE
YOLOv8 does not take torch tensor as input in python script

### ISSUE THREAD

**Issue by toschi23** [ultralytics/ultralytics#1218]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

My question is quite basic, which is why I was curious as to why I could not find an answer myself. If this is a dublicate, excuse me. 

I have a custom neural network where I want to use the output of a YOLO model run on the same input as part of my target. (E.g. number of objects in the image.)
For that I build my own class as follows:

```
from ultralytics import YOLO
class yolo_mask_model(pl.LightningModule):
	def __init__(self, weight_path = '/mypath/weights/best.pt'):
		super(yolo_mask_model, self).__init__()
		self.save_hyperparameters()
		self.pretrained_yolo = YOLO(weight_path)
		
	def forward(self, input):
		model_input = list(255 * np.transpose(input.cpu().numpy(),(0,2,3,1)))
		yolo_output = self.pretrained_yolo(model_input, stream=False)
        ... some more code
```

However I do find the step to go from `torch.tensor` to a list of numpy arrays very inefficient since it requires shifting data from my gpu to the cpu and back.

When I just plug in the tensor directly I receive the following errormessage:

`Exception has occurred: AssertionError
Expected PIL/np.ndarray image type, but got <class 'torch.Tensor'>`

I looked in the code a bit and found out, there is a lot of preprocessing done before turning the source into a tensor and running it through the model architecture. 
I could not find a trivial way, to skip all these preprocessing steps.

(I posted this question to stack overflow a week ago [here](https://stackoverflow.com/questions/75580508/yolov8-does-not-take-torch-tensor-as-input-in-python-script), but only got the answer, that  it does not work.

Is there any way around this behaviour?

**Comment by glenn-jocher** [comment]
thanks for the bug report, I'm able to reproduce using your code. 

There's no quick fix for this, as Tensor support is missing currently. In the meantime you can try a different source, i.e. a numpy array is probably closest to a torch tensor. The input formats supported are here:

https://docs.ultralytics.com/predict/


| source     | stream  | model(arg)                                 | type           | notes            |
|------------|---------|--------------------------------------------|----------------|------------------|
| image      |         | `'im.jpg'`                                 | `str`, `Path`  |                  |
| URL        |         | `'https://ultralytics.com/images/bus.jpg'` | `str`          |                  |
| screenshot |         | `'screen'`                                 | `str`          |                  |
| PIL        |         | `Image.open('im.jpg')`                     | `PIL.Image`    | HWC, RGB         |
| OpenCV     |         | `cv2.imread('im.jpg')[:,:,::-1]`           | `np.ndarray`   | HWC, BGR to RGB  |
| numpy      |         | `np.zeros((640,1280,3))`                   | `np.ndarray`   | HWC              |
| torch      |         | `torch.zeros(16,3,320,640)`                | `torch.Tensor` | BCHW, RGB        |
| CSV        |         | `'sources.csv'`                            | `str`, `Path`  | RTSP, RTMP, HTTP |         
| video      | &check; | `'vid.mp4'`                                | `str`, `Path`  |                  |
| directory  | &check; | `'path/'`                                  | `str`, `Path`  |                  |
| glob       | &check; | `path/*.jpg'`                              | `str`          | Use `*` operator |
| YouTube    | &check; | `'https://youtu.be/Zgi9g1ksQHc'`           | `str`          |                  |
| stream     | &check; | `'rtsp://example.com/media.mp4'`           | `str`          | RTSP, RTMP, HTTP |

**Comment by toschi23** [comment]
Thank you for the reply. As you can see from my code snipped converting the tensor to numpy is what I do. The main problem i have with this approach is, that this requires me to load all the tensors from my gpu to my cpu only to run the model on the gpu again, after the preprocessing. This seems to be a very unnecessary step. 

Are you aware of anybody else facing this issue? This might be a good feature to have.

**Comment by Mohit-robo** [comment]
> @toschi23 thanks for the bug report, I'm able to reproduce using your code.
> 
> There's no quick fix for this, as Tensor support is missing currently. In the meantime you can try a different source, i.e. a numpy array is probably closest to a torch tensor. The input formats supported are here:
> 
> https://docs.ultralytics.com/predict/
> 
> ```python
> from ultralytics import YOLO
> 
> model = YOLO('yolov8n.pt)
> results = model(source)
> ```
> 
> source 	stream 	model(arg) 	type 	notes
> image 		`'im.jpg'` 	`str`, `Path` 	
> URL 		`'https://ultralytics.com/images/bus.jpg'` 	`str` 	
> screenshot 		`'screen'` 	`str` 	
> PIL 		`Image.open('im.jpg')` 	`PIL.Image` 	HWC, RGB
> OpenCV 		`cv2.imread('im.jpg')[:,:,::-1]` 	`np.ndarray` 	HWC, BGR to RGB
> numpy 		`np.zeros((640,1280,3))` 	`np.ndarray` 	HWC
> torch 		`torch.zeros(16,3,320,640)` 	`torch.Tensor` 	BCHW, RGB
> CSV 		`'sources.csv'` 	`str`, `Path` 	RTSP, RTMP, HTTP
> video 	✓ 	`'vid.mp4'` 	`str`, `Path` 	
> directory 	✓ 	`'path/'` 	`str`, `Path` 	
> glob 	✓ 	`path/*.jpg'` 	`str` 	Use `*` operator
> YouTube 	✓ 	`'https://youtu.be/Zgi9g1ksQHc'` 	`str` 	
> stream 	✓ 	`'rtsp://example.com/media.mp4'` 	`str` 	RTSP, RTMP, HTTP

As mentioned here, `torch.Tensor` is accepted, so the Pytorch data loader's input should be accepted. But when I try to give input to the YOLOv8 segmentation model, it raises 
`cv2.error: OpenCV(4.8.0) /io/opencv/modules/imgproc/src/resize.cpp:3699: error: (-215:Assertion failed) !dsize.empty() in function 'resize'` 
my input image size is  `torch.Size([128, 3, 384, 128])`

**Comment by glenn-jocher** [comment]
@Mohit-robo thank you for your question! 

While it is mentioned in the documentation that `torch.Tensor` is an accepted input format for YOLOv8, it seems that there is currently an issue when directly passing a `torch.Tensor` to the YOLOv8 segmentation model. This is likely due to an assertion failure in the resizing function.

To work around this issue, you can try converting the `torch.Tensor` to a numpy array before passing it to the model. Numpy arrays are closely compatible with torch tensors and are accepted as an input format by YOLOv8. You can do this using `tensor.numpy()`.

Please let me know if you have any further questions or need additional assistance.

**Comment by MOXHAN** [comment]
Hi @glenn-jocher ,

Im trying to write a script which uses jetson-utils library for video input and output to run inference on yolov8 and show the results. So far everything seems to work, but somehow the Yolov8.engine doesnt make any detections. I did check the model on example pictures with the CLI command yolo predict and everything was fine, I got detections. But for the same image in my script it didnt work. After reading this thread, I started to think that my images being torch.tensors on the GPU might be a problem, so I converted them to numpy arrays with img.cpu().numpy() like you suggested. But now Im getting the assertion error:
`cv2.error: OpenCV(4.5.0) /opt/opencv/modules/imgproc/src/resize.cpp:3688: error: (-215:Assertion failed) !dsize.empty() in function 'resize'`

I did verify the input images by saving them and they look fine, I checked the model with the CLI as I mentioned and the Input has the right shape (1, 3, 640, 640), which Yolo also seems to accept, because I get this terminal posts:
`0: 640x640 (no detections), 4.1ms
Speed: 1.2ms preprocess, 4.1ms inference, 4.5ms postprocess per image at shape (1, 3, 640, 640)`

This is the inference part of my Script:
```
input = videoSource("/dev/video0")
# set gpu as device
torch.cuda.set_device(0)
# load model engine
model = YOLO("models/yolov8n.engine", task="detect")
# capture frames until end-of-stream (or the user exits)
while True:    
	# get Frame from Source as cuda Image on GPU
	cuda_img = input.Capture(format='rgb32f', timeout=1000)
	
	if cuda_img is None:  # if a timeout occurred
		continue
        
        
	# Resize the image to match YOLOv8's input shape (640x640)
	cuda_resized = cudaAllocMapped(width=640, height=640, format=cuda_img.format)
	cudaResize(cuda_img, cuda_resized)
	
	# Normalize for yolov8 input
	cudaNormalize(cuda_resized, (0,255), cuda_resized, (0,1))
	
	# Use cudaImage array Interface for conversion to torch.tensor
	img = torch.as_tensor(cuda_resized, device="cuda")
	
	# Convert type to float32
	img = img.float()
	# Transpose to match yolov8 input shape (3, 640, 640)
	img = img.permute(2, 0, 1)
	# Add extra dimension to represent batch size (1, 3, 640, 640)
	img = img.unsqueeze(0)
        # Move to cpu and convert to numpy array
        img = img.cpu().numpy()
        # inference
        results = model.predict(img)
```

**Comment by issahammoud** [comment]
It is frustrating to not be able to input a tensor to YOLO. If I am using it in a pipeline, I need to move the data back and forth between CPU and GPU to just do a prediction. All the speed of YOLO is gone.

Could you point out what is the problem so may we help in fixing it?

**Comment by glenn-jocher** [comment]
Thank you for your feedback. The current limitation is due to preprocessing requirements. We're working on improving this, and contributions are welcome. Please ensure you're using the latest version, as updates may address this issue.

**Comment by andreacelani** [comment]
Hi @glenn-jocher , same here. YOLOv8 fed with `numpy.ndarray` from CPU gives correct outputs (instance segmentation with boxes, masks, etc..) .
If I fed a pytorch tensor from GPU (the input coming from Zed2i stereo cam) resized and formatted as needed: BCHW, RGB  ex.: [1, 3, 384, 640] I give no output at all, boxes: None, masks: None .
As cited previously the backup solution is to move pytorch tensor to CPU with `.cpu().numpy()`  but it's basically a waste of time shifting data back and forth from GPU to CPU while all the computation could be handled by GPU only.
In my case the model is a .pt model .

**Comment by andreacelani** [comment]
Hi @glenn-jocher , actually it works. At least with .pt model . I think I fed the wrong pixel order because I never had a tensor shape error.
This is the code that works to reshape the torch tensor image in GPU [720,1280,3] to input compliant with YOLOv8 spec.:
(not so far from @MOXHAN code)
```
import torch
from torch import cuda
from torchvision.transforms import v2

...

#read from cupy ndarray in GPU
torch_image = torch.from_dlpack(self.zed_image_without_alpha)

#from 0-255 color to 0.0-1.0 values
torch_image_float = torch_image.float()/255.0

# Transpose to match yolov8 input shape (C,H,W) from (H,W,C)
torch_image_float = torch_image_float.permute(2, 0, 1)

# Add extra dimension to represent batch size (B,C,H,W) (1, 3, 720, 1280)
torch_image_float = torch_image_float.unsqueeze(0)

#resize and pad to match 32 stride and trained model (1, 3, 384, 640)
torch_image_float_resize = v2.Resize(size=(360, 640))(torch_image_float)
torch_image_padded = v2.Pad(padding=(0, 12, 0, 12))(torch_image_float_resize)

#run YoloV8 and read results
results = self.yolo.predict(source=torch_image_padded, verbose=False, stream=False, conf=self.threshold, device=self.device)
results: Results = results[0]
```

**Comment by glenn-jocher** [comment]
Great to hear it's working with the `.pt` model! It seems the issue was with the pixel order. If you have any more questions or need further assistance, feel free to ask.

**Comment by zhluo94** [comment]
Hi @glenn-jocher ,

I am running into an issue that seems similar to @MOXHAN 's above. I have got tensor inputs to correct work with .pt model, with the preprocessing code pretty much the same as @andreacelani's, except that I am using a batch size of 8. However, after switching to the TensorRT model, the model returns no detection (but no error either) and `model.device` returns `None`, both matched with @MOXHAN 's description. I exported the engine with the command:
```
yolo_model = YOLO("yolo11n.pt").to('cuda')
yolo_model.export(format="engine", batch=batch_size, dynamic = True, device = 0) 
```
 the export succeeded with the following log: 

```
Exporting to TensorRT engine
Ultralytics 8.3.58 🚀 Python-3.10.15 torch-2.4.0+cu124 CUDA:0 (Tesla T4, 14918MiB)
YOLO11n summary (fused): 238 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

PyTorch: starting from 'yolo11n.pt' with input shape (8, 3, 640, 640) BCHW and output shape(s) (8, 84, 8400) (5.4 MB)

ONNX: starting export with onnx 1.17.0 opset 19...
ONNX: slimming with onnxslim 0.1.47...
ONNX: export success ✅ 17.2s, saved as 'yolo11n.onnx' (10.1 MB)

TensorRT: starting export with TensorRT 10.7.0...
[01/23/2025-18:09:25] [TRT] [I] [MemUsageChange] Init CUDA: CPU -2, GPU +0, now: CPU 852, GPU 1312 (MiB)
[01/23/2025-18:09:27] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +966, GPU +194, now: CPU 1684, GPU 1506 (MiB)
[01/23/2025-18:09:27] [TRT] [I] ----------------------------------------------------------------
[01/23/2025-18:09:27] [TRT] [I] Input filename:   yolo11n.onnx
[01/23/2025-18:09:27] [TRT] [I] ONNX IR version:  0.0.9
[01/23/2025-18:09:27] [TRT] [I] Opset version:    19
[01/23/2025-18:09:27] [TRT] [I] Producer name:    pytorch
[01/23/2025-18:09:27] [TRT] [I] Producer version: 2.4.0
[01/23/2025-18:09:27] [TRT] [I] Domain:           
[01/23/2025-18:09:27] [TRT] [I] Model version:    0
[01/23/2025-18:09:27] [TRT] [I] Doc string:       
[01/23/2025-18:09:27] [TRT] [I] ----------------------------------------------------------------
TensorRT: input "images" with shape(-1, 3, -1, -1) DataType.FLOAT
TensorRT: output "output0" with shape(-1, 84, -1) DataType.FLOAT
TensorRT: building FP32 engine as yolo11n.engine
[01/23/2025-18:09:27] [TRT] [I] BuilderFlag::kTF32 is set but hardware does not support TF32. Disabling TF32.
[01/23/2025-18:09:27] [TRT] [I] BuilderFlag::kTF32 is set but hardware does not support TF32. Disabling TF32.
[01/23/2025-18:09:27] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
[01/23/2025-18:10:10] [TRT] [I] Compiler backend is used during engine build.
[01/23/2025-18:10:45] [TRT] [I] Detected 1 inputs and 1 output network tensors.
[01/23/2025-18:10:46] [TRT] [I] Total Host Persistent Memory: 393376 bytes
[01/23/2025-18:10:46] [TRT] [I] Total Device Persistent Memory: 1530368 bytes
[01/23/2025-18:10:46] [TRT] [I] Max Scratch Memory: 22937600 bytes
[01/23/2025-18:10:46] [TRT] [I] [BlockAssignment] Started assigning block shifts. This will take 180 steps to complete.
[01/23/2025-18:10:46] [TRT] [I] [BlockAssignment] Algorithm ShiftNTopDown took 13.2993ms to assign 10 blocks to 180 nodes requiring 154010112 bytes.
[01/23/2025-18:10:46] [TRT] [I] Total Activation Memory: 154009600 bytes
[01/23/2025-18:10:46] [TRT] [I] Total Weights Memory: 12989320 bytes
[01/23/2025-18:10:46] [TRT] [I] Compiler backend is used during engine execution.
[01/23/2025-18:10:46] [TRT] [I] Engine generation completed in 79.1343 seconds.
[01/23/2025-18:10:46] [TRT] [I] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 1 MiB, GPU 261 MiB
TensorRT: export success ✅ 98.7s, saved as 'yolo11n.engine' (13.8 MB)
```

I assume the export was successful. I then loaded the engine with `yolo_model = YOLO("yolo11n.engine", task="detect")` ([as suggested here](https://github.com/ultralytics/ultralytics/issues/7644#issuecomment-1898633422)) (some error showed up if `to("cuda")` is added). I also confirmed that the autobackend.py correctly set the input tensor shape for the prediction. I am now not sure what may go wrong and how I should proceed. Any suggestion will be highly appreciated! Thanks!

**Comment by glenn-jocher** [comment]
Thank you for the detailed report. The issue may stem from improperly configured dynamic shapes during export. Ensure `dynamic=True` was correctly set during export, and confirm the input tensor shape matches the exported model's dynamic range. Additionally, verify your TensorRT engine is correctly loaded and configured on a supported GPU environment. For further debugging, consider testing with a simpler input (e.g., single image with batch size 1) to isolate potential causes. Let me know if the issue persists.
